api/spanner/push_run.go

```diff
@@ -73,38 +304,294 @@ func pushRun(ctx context.Context, api API, id int64) {
+
+	for _, result := range report.Results {
+		t := NewTest(result.Test, nil)
+		ss.AddTest(t)
+		res := NewResult(result.Status, nil)
+		ss.AddResult(res)
+		ss.AddRunResult(r, res)
+		ss.AddRunResultTest(r, res, t, result.Message)
+
+		names := mapset.NewSet()
+		for _, s := range result.Subtests {
+			if names.Contains(s.Name) {
+				logger.Warningf("Found test \"%s\" contains duplicate subtest name \"%s\"", result.Test, s.Name)
+			}
+			names.Add(s.Name)
+
+			t := NewTest(result.Test, &s.Name)
+			ss.AddTest(t)
+			res := NewResult(s.Status, nil)
+			ss.AddResult(res)
+			ss.AddRunResult(r, res)
+			ss.AddRunResultTest(r, res, t, s.Message)
+		}
+	}
+	report = nil
+
+	s := semaphore.NewWeighted(numConcurrentBatches)
 
-	n, err := numRowsToUpload(ctx, sClient, id, report)
+	logger.Infof("Updating tests from %d-test run %d", len(ss.Tests), run.ID)
+
+	fatalLock := &sync.Mutex{}
+	fatalErrs := make([]error, 0)
+	reportFatal := func(err error) {
+		fatalLock.Lock()
+		defer fatalLock.Unlock()
+		fatalErrs = append(fatalErrs, err)
+	}
+	hasFatal := func() bool {
+		fatalLock.Lock()
+		defer fatalLock.Unlock()
+		return len(fatalErrs) > 0
+	}
+
+	writeTestBatch := func(batchSync *semaphore.Weighted, rowGroupSync *sync.WaitGroup, rows []*spanner.Mutation, tests []*Test, m, n int) {
+		defer rowGroupSync.Done()
+		defer batchSync.Release(1)
+		batch := rows[m:n]
+		ts := tests[m:n]
+
+		err := retry.Do(func() error {
+			logger.Infof("Writing test batch for %d-test run %d: [%d,%d)", len(rows), run.ID, m, n)
+
+			for i := range batch {
+				err := func(ins *spanner.Mutation, t *Test) error {
+					insCtx, insCancel := context.WithTimeout(ctx, time.Second*60)
+					defer insCancel()
+
+					_, err = sClient.Apply(insCtx, []*spanner.Mutation{ins})
+					if err == nil {
+						logger.Infof("Wrote new test to Spanner: %v", *t)
+						return nil
+					}
+
+					// Continue with hash collission check iff error was Cloud Spanner error
+					// "already exists". Other errors are unexpected.
+					spanErr, ok := err.(*spanner.Error)
+					if !ok || spanErr.Code != codes.AlreadyExists {
+						return err
+					}
+
+					readCtx, readCancel := context.WithTimeout(ctx, time.Second*60)
+					defer readCancel()
+
+					row, err := sClient.Single().ReadRow(readCtx, testsTableName, spanner.Key{t.TestID, t.SubtestID}, []string{testNameColumnName, subtestNameColumnName})
+					if err != nil {
+						return err
+					}
+
+					var testName string
+					err = row.ColumnByName(testNameColumnName, &testName)
+					if err != nil {
+						return err
+					}
+					var subtestName spanner.NullString
+					err = row.ColumnByName(subtestNameColumnName, &subtestName)
+					if err != nil {
+						return err
+					}
+					if t.TestName != testName || t.SubtestName != subtestName {
+						return fmt.Errorf(`Hash collision: Test identifier <%d, %v> mapped to different test+subtest names: "%s".%v != "%s".%v`, t.TestID, t.SubtestID, t.TestName, t.SubtestName, testName, subtestName)
```

Comment:
@lukebjerring do you think we should accumulate these collisions, remove them from the result set, then push the nearly-complete run, rather than giving up entirely?

Reply:
I also concur with Luke here. It's fine to lose a few lines, and we still want the otherwise nearly complete run. We just need to log the error here. Basically, I agree with you this is unlikely to happen, so we shouldn't worry about it prematurely, but we do want to know when/if it happens.